tracer/src/Datadog.Trace/Debugger/Instrumentation/AsyncMethodDebuggerInvoker.cs

```diff
@@ -0,0 +1,367 @@
+        {
+            if (ProbeRateLimiter.Instance.IsLimitReached)
+            {
+                return AsyncMethodDebuggerState.CreateInvalidatedDebuggerState();
+            }
+
+            if (isReEntryToMoveNext)
+            {
+                return AsyncContext.Value;
+            }
+
+            isReEntryToMoveNext = true;
+            var kickoffInfo = AsyncHelper.GetAsyncKickoffMethodInfo(instance);
+            if (kickoffInfo.KickoffParentObject == null && kickoffInfo.KickoffMethod.IsStatic == false)
+            {
+                Log.Error($"{nameof(BeginMethod)}: hoisted 'this' has not found. {kickoffInfo.KickoffParentType.Name}.{kickoffInfo.KickoffMethod.Name}");
+            }
+
+            var asyncState = SetContext(kickoffInfo);
+
+            if (!MethodMetadataProvider.TryCreateIfNotExists(instance, methodMetadataIndex, in methodHandle, in typeHandle, kickoffInfo))
+            {
+                Log.Warning($"BeginMethod_StartMarker: Failed to receive the InstrumentedMethodInfo associated with the executing method. type = {typeof(TTarget)}, instance type name = {instance?.GetType().Name}, methodMetadataId = {methodMetadataIndex}");
+                return AsyncMethodDebuggerState.CreateInvalidatedDebuggerState();
+            }
+
+            // We are not in a continuation,
+            // i.e. there is no context or the context just created in the kick off method but the asyncState machine not yet run
+            // so create a new one and capture everything
+            asyncState.ProbeId = probeId;
+            asyncState.StartTime = DateTimeOffset.UtcNow;
+            asyncState.MethodMetadataIndex = methodMetadataIndex;
+            asyncState.MoveNextInvocationTarget = instance;
+
+            BeginMethodStartMarker(ref asyncState);
+
+            BeginMethodLogArgs(instance, ref asyncState);
+
+            asyncState = BeginMethodEndMarker(ref asyncState);
+
+            AsyncContext.Value = asyncState;
+            return AsyncContext.Value;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void BeginMethodStartMarker(ref AsyncMethodDebuggerState asyncState)
+        {
+            asyncState.SnapshotCreator.StartDebugger();
+            asyncState.SnapshotCreator.StartSnapshot();
+            asyncState.SnapshotCreator.StartCaptures();
+            asyncState.SnapshotCreator.StartEntry();
+            asyncState.SnapshotCreator.CaptureInstance(asyncState.KickoffInvocationTarget, asyncState.KickoffInvocationTargetType);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void BeginMethodLogArgs<TTarget>(TTarget instance, ref AsyncMethodDebuggerState asyncState)
+        {
+            // capture hoisted arguments
+            // todo: should we save it in the state for later use?
```

Comment:
I agree, I think we should keep aside all the `FieldInfo` (inside the state machine) that we determined are hoisted parameters, so we won't pay for the calls to `instance.GetType().GetFields`/`.GetParameters` etc more than once.
(Same for `GetHoistedLocalsFromStateMachine`).

Reply:
Fixed in 871a652d9b150d63443d87a542e34d4643145adf